src/main/java/faang/school/projectservice/service/StageService.java

```diff
@@ -0,0 +1,123 @@
+import java.util.List;
+
+@Component
+@RequiredArgsConstructor
+public class StageService {
+    private final StageRepository stageRepository;
+    private final ProjectRepository projectRepository;
+    private final StageMapper stageMapper;
+    public void makeStage(String stageName, Long projectId, List<StageRoles> stageRoles) {
+        List<Task> tasks = new ArrayList<>();
+        List<TeamMember> executors = new ArrayList<>();
+        makeStage(stageName, projectId, stageRoles, tasks, executors);
+    }
+
+    public void makeStage(String stageName, Long projectId, List<StageRoles> stageRoles, List<Task> tasks, List<TeamMember> executors) {
+        if (projectRepository.getProjectById(projectId).getStatus() == ProjectStatus.CANCELLED) {
+            throw new ProjectStatusException("You can not make stage for cancelled project!");
+        }
+        if (projectRepository.getProjectById(projectId).getStatus() == ProjectStatus.COMPLETED) {
+            throw new ProjectStatusException("You can not make stage for completed project!");
+        }
+        Stage stage = new Stage();
+        stage.setStageName(stageName);
+        stage.setProject(projectRepository.getProjectById(projectId));
+        stage.setStageRoles(stageRoles);
+        stage.setTasks(tasks);
+        stage.setExecutors(executors);
+        stageRepository.save(stage);
+    }
+
+    public List<StageDto> getStagesByStatus(Long projectId, TaskStatus status) {
+        List<Stage> projectStages = stageRepository.findAll().stream()
+                .filter(stage -> stage.getProject().getId() == projectId).toList();
+        return stageMapper.toDto(projectStages.stream()
+                .filter(stage -> {
+                    if (stage.getTasks().stream().filter(task -> task.getStatus().equals(status)).toList().isEmpty()) {
+                        return false;
+                    }
+                    return true;
+                })
+                        .toList());
+    }
+
+    public void deleteStage(Long stageId) {
+        Stage stage = stageRepository.getById(stageId);
+        List<Task> tasks = stage.getTasks();
+        for (Task task : tasks) {
+            task.setStatus(TaskStatus.CANCELLED);
+        }
+        stageRepository.delete(stage);
+    }
+
+    public void deleteStage(Long stageIdToDelete, Long stageIdToReceive) {
+        Stage stageToDelete = stageRepository.getById(stageIdToDelete);
+        Stage stageToReceive = stageRepository.getById(stageIdToReceive);
+        List<Task> tasks = stageToDelete.getTasks();
+        for (Task task : tasks) {
+            task.setStage(stageToReceive);
+        }
+        stageRepository.delete(stageToDelete);
+    }
+
+    public void updateStage(Long stageId) {
+        Stage stage = stageRepository.getById(stageId);
+        Project project = stage.getProject();
+        List<TeamMember> executors = stage.getExecutors();
+        for (StageRoles stageRole : stage.getStageRoles()) {
```

Comment:
consider to break it down to different methods or even make a separate class for this logic

Reply:
make a different method